diag_manager/fms_diag_object.F90

```diff
@@ -777,39 +784,54 @@ subroutine fms_diag_do_io(this, is_end_of_run)
       call diag_file%write_axis_data(this%diag_axis)
     endif
 
-    is_writing = diag_file%is_time_to_write(model_time)
+    is_writing = .false.
 
     ! finish reduction method if its time to write
-    buff_reduct: if (is_writing) then
-      buff_ids = diag_file%FMS_diag_file%get_buffer_ids()
-      ! loop through the buffers and finish reduction if needed
-      buff_loop: do ibuff=1, SIZE(buff_ids)
-        diag_buff => this%FMS_diag_output_buffers(buff_ids(ibuff))
-        field_yaml => diag_yaml%diag_fields(diag_buff%get_yaml_id())
-        diag_field => this%FMS_diag_fields(diag_buff%get_field_id())
-
-        ! Go away if there is no data to write
-        if (.not. diag_buff%is_there_data_to_write()) cycle
-
-        ! sets missing value
-        mval = diag_field%find_missing_value(missing_val)
-        ! time_average and greater values all involve averaging so need to be "finished" before written
-        if( field_yaml%has_var_reduction()) then
-          if( field_yaml%get_var_reduction() .ge. time_average) then
-            if(DEBUG_REDUCT)call mpp_error(NOTE, "fms_diag_do_io:: finishing reduction for "//diag_field%get_longname())
-            has_mask = diag_field%has_mask_variant()
-            if(has_mask) has_mask = diag_field%get_mask_variant()
-            error_string = diag_buff%diag_reduction_done_wrapper( &
+    buff_ids = diag_file%FMS_diag_file%get_buffer_ids()
+
+    ! loop through the buffers and finish reduction if needed
+    buff_loop: do ibuff=1, SIZE(buff_ids)
+      diag_buff => this%FMS_diag_output_buffers(buff_ids(ibuff))
+      field_yaml => diag_yaml%diag_fields(diag_buff%get_yaml_id())
+      diag_field => this%FMS_diag_fields(diag_buff%get_field_id())
+
+      ! Go away if there is no data to write
+      if (.not. diag_buff%is_there_data_to_write()) cycle
+
+      ! Go away if not time to write for this field
+      ! otherwise continue and write after this loop
+      if(.not. diag_file%is_time_to_write(diag_field%current_model_time)) then
+        cycle
+      else
+        is_writing = .true.
+      endif
+
+      ! all fields should have the same time if writing
+      if(ibuff .eq. 1 ) &
+        model_time = diag_field%current_model_time
+      if(diag_field%current_model_time .ne. model_time) then
+        call mpp_error(FATAL, "fms_diag_do_io:: each field's model time should be the same for a given file")
```

Comment:
This is not always the case. There are many different time steps:
dt_atmos = 1800, dt_cpld  = 7200, rad_time_step= 10800, sw_rad_time_step = 3600

Reply:
I'll take that out, but I'm not sure what that means for the `is_writing` if block below since its using the model time to update the file objects:
 https://github.com/rem1776/FMS/blob/9809760ce34b8d1bb6f8b81fc1b9552e49b1c13e/diag_manager/fms_diag_object.F90#L837-L842

I was thinking of adding a loop for each field's current model time (or just each unique model time), just not sure if that would mess with the write routines.